ast/ast.hpp

```diff
@@ -210,13 +217,12 @@ struct ModifiablePrimary : Expression {
     }
     void accept(Visitor& v) override { v.visit(this); }
 };
-struct RoutineCall : Expression {
+struct RoutineCall : Expression, Statement {
     lexer::Token routine;
     std::vector<sPtr<Expression>> args;
 
     bool operator==(const RoutineCall& other) const {
-        return Node::operator==(other) && args == other.args &&
-               routine == other.routine;
+        return args == other.args && routine == other.routine;
```

Comment:
Where did `Node::operator==(other)` go? Was it causing an error or what?

Reply:
Compilation error due to the fact, that `Node` is root for both "root" classes: `Expression` and `Statement`

Reply:
Hmm, interesting. I cannot find an explanation for that online. Maybe @kureduro can help?
Actually, I think it makes more sense to call the `==` operator of both parents rather than just the common ancestor.

Reply:
> Hmm, interesting. I cannot find an explanation for that online. Maybe @kureduro can help?
> Actually, I think it makes more sense to call the `==` operator of both parents rather than just the common ancestor.

http://eel.is/c++draft/class.mi#3
The equality operator is just a regular method, and since both `Expression` and `Statement` inherit from `Node`, there's an ambiguity, since there are two different `Node` classes to choose the `Node::operator==` from.

Reply:
But why doesn't it detect that both `Node` classes are actually one and the same?
Maybe C++ is just not as good as Eiffel when it comes to multiple inheritance)